crates/common/src/compile.rs

```diff
@@ -298,6 +299,7 @@ impl ProjectCompiler {
     fn zksync_compile_with<F>(
         self,
         root_path: impl AsRef<Path>,
+        version: &Version,
```

Comment:
Not sure that we want to fix this here. I believe in solc/vyper the version logging is done inside compilers, so the more we can mirror that the better. For us it is a bit broken because for a given Compiler, the specified version is both used to group sources to be compiled and reporting of that specific version when logging. In our case we need to use the solc version for the former and zksolc version we pretty much don't care about inside compilers for now(except for logging and caching), hence why it spits out the solc vs the zksolc version on the output. Imo once everything works we should look at the best way to adapt things to fix the reporting of the version there, and how to represent zksolc/solc pair for a compiler instance correctly so everything fits. If we need to debug the zksolc version we can use `trace` log level. If this makes life easier maybe a quick fix [here](https://github.com/Moonsong-Labs/compilers/blob/a472fde21e0232f07d803d517b29a9bf92da9f95/crates/compilers/src/zksync/compile/project.rs#L453) can be done instead?

Reply:
I propose we use some part of the BuildMetadata to encode one either of the versions, which is already in use for wrapper projects to keep the project's version and the wrapper's version separate yet explicit

Reply:
This version actually comes from 

```bash
$ ./zksolc-linux-amd64-musl-v1.5.0 --version
Solidity compiler for ZKsync v1.5.0 (LLVM build f31171e1501f84ec863439d402b10e829ca19d26)
```

whereas the solidity version is:
```bash
$ ./solc-linux-amd64-0.8.20-1.0.1 --version
solc, the ZKsync Solidity compiler commandline interface
Version: 0.8.20+commit.9f2a5241.Linux.clang
ZKsync: 0.8.20-1.0.1
```

The `version` here is already present in compilers, and was fixed in https://github.com/Moonsong-Labs/compilers/commit/a472fde21e0232f07d803d517b29a9bf92da9f95 which was using `nth(3)` instead of `nth(4)` thereby failing the version parsing. As such, this PR doesn't really introduce a new paradigm, but merely prints out _what little_ information we already have insofar as a sanity check.

Reply:
My point is more to do things as close as possible to upstream where a) `compile_with` takes no version, b) That line is logged inside foundry compilers, where we should log this instead imo. Not a big deal either way though.
As regards version parsing, if we want to support using versions previous to 1.5.0, we might want to check it parses previous versions correctly

Reply:
I agree, could you point to where this info can be logged instead? IIRC https://github.com/Moonsong-Labs/compilers/blob/a472fde21e0232f07d803d517b29a9bf92da9f95/crates/compilers/src/zksync/compile/project.rs#L453 gives the solidity version and _not_ the zksolc version.

Reply:
Perhaps we can print this here for now? https://github.com/Moonsong-Labs/compilers/blob/a472fde21e0232f07d803d517b29a9bf92da9f95/crates/compilers/src/zksync/compile/project.rs#L353

Reply:
As a quick and dirty thing. Inside `compile_sequential` you do have a `ZkSolc` instance that you can get the version from: https://github.com/Moonsong-Labs/compilers/blob/a472fde21e0232f07d803d517b29a9bf92da9f95/crates/compilers/src/zksync/compile/project.rs#L442.
Final solution would probably be to add zksolc version as a parameter to ZkSolcVersionedInput and produce a version that has either zksolc or solc version to the `BuildMetadata` (as @Karrq pointed out) so that it outputs both versions being used as a result of the `version()` call, but we have to think about it a bit because that is also used for version resolution (what solc to use to compile each source)

Reply:
I decided to localize the version for now